api/services/Reddit.js

```diff
@@ -65,6 +68,18 @@ var makeRequest = async function (refreshToken, requestType, url, data, rateLimi
   return bodyJson;
 };
 
+var getEntireListing = async function (refreshToken, endpoint, query, rateThreshold, after, before) {
+  var url = endpoint + query + (query ? '&' : '?') + 'count=102&limit=100' + (after ? '&after=' + after : '') + (before ? '&before=' + before : '');
+  var batch = await makeRequest(refreshToken, 'GET', url, undefined, rateThreshold, after, before);
+  var results = batch.data.children;
+  if ((after || after === undefined) && batch.data.after === null || before && batch.data.before === null) {
```

Comment:
You're right, I think there might have been a logic error, so I simplified the logic a bit. It's still a bit complicated, so I'll explain the process.

`getEntireListing()` can be called in one of the following three states:
Case 1. `after` is defined and `before` is not (i.e. we're searching forward, getting all listings after a certain point)
Case 2. `before` is defined and `after` is not (i.e. we're searching backward, getting all listings before a certain point)
Case 3. `after` and `before` are both `undefined` (i.e. we're searching forward, getting the entire listing starting at the top)

Whenever we make a call, reddit returns `batch.data.after` and `batch.data.before`, which are intended to be starting points for the next search. However, one of these will be `null` if there are no more results to get. We want to recursively call `getEntireListing()` again with these new parameters as long as they're not `null`.

Example: Suppose there are 250 modmails in total, and we want to get all of them. 

1. We call `getEntireListing()` with `after` and `before` both undefined. Reddit returns a `batch` containing modmails 1-100. It also sends `batch.data.after=<id of 100th modmail>` and `batch.data.before=null`.
2. We recursively call `getEntireListing()` again, with `after=<id of 100th modmail>` and `before=undefined`. Reddit returns the batch containing modmails 101-200. It also sends `batch.data.after=<id of 200th modmail>` and `batch.data.before=<id of 101st modmail>`.
3. We recursively call `getEntireListing()` again, with `after=<id of 200th modmail>` and `before=undefined`. Reddit returns the batch containing modmails 201-250. It also sends `batch.data.after=null` (since there are no more results to get) and `batch.data.before=<id of 201st modmail>`.
4. We notice that `batch.data.after` is `null`, so we conclude that the search has finished and we return the final list.

After we get the batch, we do the following:

    after = before ? undefined : batch.data.after;
    before = before ? batch.data.before : undefined;

In other words:

* If `before` is undefined (i.e. we're searching forward), set `after` to `batch.data.after` and keep `before` as `undefined` for the next search.
* If `before` is defined (i.e. we're searching backward), set `after` to `undefined` and set `before` to `batch.data.before` for the next search.

Afterwards, test `if (!before && !after)`. One of these will always be undefined, and the other will be the relevant `batch.data` response from reddit. If that `batch.data` response is `null`, then the search is over; the `if` statement will check out and the results will be returned. If the `batch.data` response is something else, the if statement will not check out, and a query will be made with the new parameters.